[CMake] Clang warnings about C code being built in C++ mode

lauromoura opened the pull request:
#### e071c895f5fcd0dcbc1854af9dbdbc18883583a7
<pre>
[CMake] Clang warnings about C code being built in C++ mode
<a href="https://bugs.webkit.org/show_bug.cgi?id=245176">https://bugs.webkit.org/show_bug.cgi?id=245176</a>

Reviewed by Yusuke Suzuki.

* Source/bmalloc/CMakeLists.txt: Use a flag to control how to treat C as
  CXX.
* Source/bmalloc/libpas/src/libpas/pas_utils.h: Add ignore
  atomic-alignment in some clang atomics
* Source/cmake/WebKitCompilerFlags.cmake: Set the C_AS_CXX flag for gcc
  and clang.

Canonical link: <a href="https://commits.webkit.org/256626@main">https://commits.webkit.org/256626@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6a6ef8615d2c68f9132a9f64f627aa5a2d9a4a82

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/94492 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/3670 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/27420 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/104146 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/164416 "Built successfully and passed tests") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/3721 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/31866 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/86818 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/100119 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/100162 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/2671 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/80882 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/29705 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/84594 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/84185 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/72599 "Found 1 new API test failure: /WebKitGTK/TestWebKitAccessibility:/webkit/WebKitAccessibility/accessible/state (failure)") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/85695 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/38267 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/18009 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/80863 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/36131 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/19282 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/27795 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/4599 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/40030 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/41906 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/83530 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/41982 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/38512 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/18871 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

donny-dont commented on Source/bmalloc/CMakeLists.txt:
```diff
@@ -683,7 +683,9 @@ set_source_files_properties(
-    PROPERTIES LANGUAGE CXX
+    PROPERTIES
+    LANGUAGE CXX
+    COMPILE_OPTIONS "-xc++"
```
MSVC isn't currently building bmalloc but if it were then we probably would want to have a `COMPILE_C_AS_CXX` variable that's set based on the compiler and use that.

lauromoura replied:
Do you mean adding a `COMPILE_C_AS_CXX` guard around this block so we can turn it on/off from `WebKitCompilerFlags.cmake`?

donny-dont replied:
No I mean

```
COMPILE_OPTIONS ${COMPILE_C_AS_CXX}
```

Constellation reviewed (approved):
r=me

lauromoura committed:
[CMake] Clang warnings about C code being built in C++ mode
https://bugs.webkit.org/show_bug.cgi?id=245176

Reviewed by Yusuke Suzuki.

* Source/bmalloc/CMakeLists.txt: Use a flag to control how to treat C as
  CXX.
* Source/bmalloc/libpas/src/libpas/pas_utils.h: Add ignore
  atomic-alignment in some clang atomics
* Source/cmake/WebKitCompilerFlags.cmake: Set the C_AS_CXX flag for gcc
  and clang.

Canonical link: https://commits.webkit.org/256626@main
